strangelove-ventures/packet-forward-middleware
# refactor: add new metadata fields for fine tuned behavior in the middleware

While finishing the implementation of swap and forward a few edge cases were discovered that require some way to pass signals between various middleware in a stack.

This PR adds three new fields to the `ForwardMetadata`
- `Nonrefundable`
- `Processed`
- `DisableDenomComposition`

`Nonrefundable` is used to signal to the packet forward middleware that some action has occurred making a packet non-refundable e.g. a swap occurs and can't be undone so funds should remain on the intermediate chain where the swap happened if the forward fails.

`Processed` is used to signal to the packet forward middleware that some other middleware in the stack has already called into the underlying application to invoke the transfer modules `OnRecvPacket` callback which means that the minting/burning of vouchers has already occurred and should not happen again.

`DisableDenomComposition` is used to signal to the packet forward middleware that some token being forwarded already has a denom from the perspective of the current chain and so winding/unwinding of the denom is not necessary.

To test this:

You will need to setup heighliner and clone down this branch of our Gaia fork which imports this branch.
https://github.com/strangelove-ventures/gaia/tree/justin/forward_middleware_v3_refactor

Once that is done you can `heighliner build -c gaia --local` to build the local Docker image for Gaia

With the image built you should be able to run the updated test case in this branch
https://github.com/strangelove-ventures/ibctest/tree/justin/update-v3
